File: apps/api/app/services/rag/context_packing.py
"""Token-aware context packing for grounded prompts.

Ported from SlimX-AI ControlRoom's context_packing.py (MIT). Admits whole chunks under a token
budget (never cut mid-sentence, citation never detached from its text), caps per parent, and skips
near-identical siblings. Every chunk is recorded as admitted or rejected (with token count + reason)
so the reader can show exactly what was sent. Token counts use a dependency-free ~4-chars/token
estimate; SlimX owns model execution and post-hoc Usage.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import Any

_CHARS_PER_TOKEN = 4


def estimate_tokens(text: str) -> int:
    n = len(text or "")
    return 0 if n == 0 else max(1, math.ceil(n / _CHARS_PER_TOKEN))


@dataclass(slots=True)
class CandidateChunk:
    chunk_id: str
    text: str
    citation: str = ""
    document_id: str | None = None
    page: int | None = None
    section: str | None = None
    parent_id: str | None = None
    score: float = 0.0


@dataclass(slots=True)
class AdmittedChunk:
    chunk: CandidateChunk
    tokens: int


@dataclass(slots=True)
class PackResult:
    admitted: list[AdmittedChunk] = field(default_factory=list)
    rejected: list[dict[str, Any]] = field(default_factory=list)
    total_tokens: int = 0
    budget_tokens: int = 0


def pack_chunks(
    candidates: list[CandidateChunk], *, budget_tokens: int, max_per_parent: int = 2
) -> PackResult:
    """Admit whole chunks in order until the token budget is exhausted.

    Order is preserved (callers pass chunks already ranked by retrieval). A chunk is rejected —
    never truncated — when it would overflow the budget, when its parent is already at the cap, or
    when its text duplicates an admitted chunk.
    """
    result = PackResult(budget_tokens=max(0, budget_tokens))
    per_parent: dict[str, int] = {}
    seen: set[str] = set()
    used = 0

    def rejected(cand: CandidateChunk, reason: str, tokens: int) -> dict[str, Any]:
        return {
            "chunk_id": cand.chunk_id,
            "reason": reason,
            "tokens": tokens,
            "score": cand.score,
            "document_id": cand.document_id,
            "page": cand.page,
        }

    for cand in candidates:
        tokens = estimate_tokens(cand.text) + estimate_tokens(cand.citation)
        norm = " ".join((cand.text or "").split()).lower()
        parent = cand.parent_id or cand.chunk_id
        if not (cand.text or "").strip():
            result.rejected.append(rejected(cand, "empty", 0))
            continue
        if norm in seen:
            result.rejected.append(rejected(cand, "duplicate_text", tokens))
            continue
        if per_parent.get(parent, 0) >= max_per_parent:
            result.rejected.append(rejected(cand, "parent_cap", tokens))
            continue
        if used + tokens > result.budget_tokens:
            result.rejected.append(rejected(cand, "budget_exhausted", tokens))
            continue
        result.admitted.append(AdmittedChunk(chunk=cand, tokens=tokens))
        used += tokens
        seen.add(norm)
        per_parent[parent] = per_parent.get(parent, 0) + 1
    result.total_tokens = used
    return result

File: apps/web-lite/src/rag/contextPacking.ts
// Line-for-line port of apps/api/app/services/rag/context_packing.py. Admits whole chunks under a
// token budget (never cut mid-sentence, citation never detached), caps per parent, skips
// near-identical siblings, and records every rejection with a reason.

const CHARS_PER_TOKEN = 4;

export function estimateTokens(text: string): number {
  const n = (text || '').length;
  return n === 0 ? 0 : Math.max(1, Math.ceil(n / CHARS_PER_TOKEN));
}

export type CandidateChunk = {
  chunk_id: string;
  text: string;
  citation?: string;
  document_id?: string | null;
  page?: number | null;
  section?: string | null;
  parent_id?: string | null;
  score?: number;
};

export type AdmittedChunk = { chunk: CandidateChunk; tokens: number };

export type RejectedChunk = {
  chunk_id: string;
  reason: 'empty' | 'duplicate_text' | 'parent_cap' | 'budget_exhausted';
  tokens: number;
  score: number;
  document_id: string | null;
  page: number | null;
};

export type PackResult = {
  admitted: AdmittedChunk[];
  rejected: RejectedChunk[];
  total_tokens: number;
  budget_tokens: number;
};

export function packChunks(
  candidates: CandidateChunk[],
  { budgetTokens, maxPerParent = 2 }: { budgetTokens: number; maxPerParent?: number },
): PackResult {
  const result: PackResult = {
    admitted: [],
    rejected: [],
    total_tokens: 0,
    budget_tokens: Math.max(0, budgetTokens),
  };
  const perParent = new Map<string, number>();
  const seen = new Set<string>();
  let used = 0;

  const rejected = (
    cand: CandidateChunk,
    reason: RejectedChunk['reason'],
    tokens: number,
  ): RejectedChunk => ({
    chunk_id: cand.chunk_id,
    reason,
    tokens,
    score: cand.score ?? 0,
    document_id: cand.document_id ?? null,
    page: cand.page ?? null,
  });

  for (const cand of candidates) {
    const tokens = estimateTokens(cand.text) + estimateTokens(cand.citation ?? '');
    const norm = (cand.text || '').split(/\s+/).join(' ').trim().toLowerCase();
    const parent = cand.parent_id || cand.chunk_id;
    if (!(cand.text || '').trim()) {
      result.rejected.push(rejected(cand, 'empty', 0));
      continue;
    }
    if (seen.has(norm)) {
      result.rejected.push(rejected(cand, 'duplicate_text', tokens));
      continue;
    }
    if ((perParent.get(parent) ?? 0) >= maxPerParent) {
      result.rejected.push(rejected(cand, 'parent_cap', tokens));
      continue;
    }
    if (used + tokens > result.budget_tokens) {
      result.rejected.push(rejected(cand, 'budget_exhausted', tokens));
      continue;
    }
    result.admitted.push({ chunk: cand, tokens });
    used += tokens;
    seen.add(norm);
    perParent.set(parent, (perParent.get(parent) ?? 0) + 1);
  }
  result.total_tokens = used;
  return result;
}
